### PR TITLE
docker: fix config file

### LINKED ISSUE #3349: docker fails to start on dev variants
**Image I'm using:**
`aws-dev` from 00eef00d.

**What I expected to happen:**
Docker service to start

**What actually happened:**
Docker service failed with this error:
```
Aug 20 23:11:25 ip-10-0-60-129.us-west-2.compute.internal systemd[1]: Starting docker engine...
Aug 20 23:11:26 ip-10-0-60-129.us-west-2.compute.internal dockerd[1260]: unable to configure the Docker daemon with file /etc/docker/daemon.json: invalid character '}' looking for beginning of object key string
```

**How to reproduce the problem:**
Build and run the `aws-dev` variant.

### PR DESCRIPTION
**Pull request by vyaghras**
**Issue number:**

Closes #3349

**Description of changes:**
Handle the trailing commas in the config file template for following different combinations of settings:
- No `settings.oci-defaults`, no `settings.container-registry.mirrors`
- Only `settings.oci-defaults.capabilities`
- Only `settings.oci-defaults.capabilities` and resource-limits
- With `settings.oci-defaults` and `settings.container-registry.mirrors`

**Testing done:**
- [x] aws-dev variants 

```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },  "selinux-enabled": true
}
Aug 21 21:36:17 ip-172-31-30-186.ec2.internal dockerd[1209]: time="2023-08-21T21:36:17.020116608Z" level=info msg="Docker daemon" commit=3056208812eb5e792fa99736c9167d1e10f4ab49 graphdriver(s)=overlay2 version=20.10.21
Aug 21 21:36:17 ip-172-31-30-186.ec2.internal dockerd[1209]: time="2023-08-21T21:36:17.020247126Z" level=info msg="Daemon has completed initialization"
Aug 21 21:36:17 ip-172-31-30-186.ec2.internal systemd[1]: Started docker engine.
Aug 21 21:36:17 ip-172-31-30-186.ec2.internal dockerd[1209]: time="2023-08-21T21:36:17.060137356Z" level=info msg="API listen on /run/docker.sock"
```

- [x]  aws-ecs-1 with capabilities and rlimits
```
etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },"default-capabilities":["CAP_AUDIT_WRITE",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE",
"CAP_FOWNER",
"CAP_FSETID",
"CAP_KILL",
"CAP_MKNOD",
"CAP_NET_BIND_SERVICE",
"CAP_NET_RAW",
"CAP_SETFCAP",
"CAP_SETGID",
"CAP_SETPCAP",
"CAP_SETUID",
"CAP_SYS_CHROOT"],
"default-ulimits": { "nofile":{ "Name": "nofile", "Hard": 4096, "Soft": 1024 }}, 
"selinux-enabled": true
}
Aug 21 22:07:18 ip-172-31-19-88.ec2.internal dockerd[1697]: time="2023-08-21T22:07:18.442978887Z" level=info msg="Docker daemon" commit=3056208812eb5e792fa99736c9167d1e10f4ab49 graphdriver(s)=overlay2 version=20.10.21
Aug 21 22:07:18 ip-172-31-19-88.ec2.internal dockerd[1697]: time="2023-08-21T22:07:18.443082332Z" level=info msg="Daemon has completed initialization"
Aug 21 22:07:18 ip-172-31-19-88.ec2.internal systemd[1]: Started docker engine.
Aug 21 22:07:18 ip-172-31-19-88.ec2.internal dockerd[1697]: time="2023-08-21T22:07:18.477146046Z" level=info msg="API listen on /run/docker.sock"

```
- [x] aws-ecs-1 with capabilities, rlimits and container-registry.mirrors
```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },"default-capabilities":["CAP_AUDIT_WRITE",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE",
"CAP_FOWNER",
"CAP_FSETID",
"CAP_KILL",
"CAP_MKNOD",
"CAP_NET_BIND_SERVICE",
"CAP_NET_RAW",
"CAP_SETFCAP",
"CAP_SETGID",
"CAP_SETPCAP",
"CAP_SETUID",
"CAP_SYS_CHROOT"],"default-ulimits": { "nofile":{ "Name": "nofile", "Hard": 4096, "Soft": 1024 }},  
"registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
Aug 21 22:09:27 ip-172-31-19-88.ec2.internal dockerd[2765]: time="2023-08-21T22:09:27.755483685Z" level=info msg="Docker daemon" commit=3056208812eb5e792fa99736c9167d1e10f4ab49 graphdriver(s)=overlay2 version=20.10.21
Aug 21 22:09:27 ip-172-31-19-88.ec2.internal dockerd[2765]: time="2023-08-21T22:09:27.755536781Z" level=info msg="Daemon has completed initialization"
Aug 21 22:09:27 ip-172-31-19-88.ec2.internal systemd[1]: Started docker engine.
Aug 21 22:09:27 ip-172-31-19-88.ec2.internal dockerd[2765]: time="2023-08-21T22:09:27.811444210Z" level=info msg="API listen on /run/docker.sock"
```

- [x]  aws-ecs-1 with capabilities and rlimits
```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" }, "nvidia": { "path": "nvidia-oci" } },"default-capabilities":["CAP_AUDIT_WRITE",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE",
"CAP_FOWNER",
"CAP_FSETID",
"CAP_KILL",
"CAP_MKNOD",
"CAP_NET_BIND_SERVICE",
"CAP_NET_RAW",
"CAP_SETFCAP",
"CAP_SETGID",
"CAP_SETPCAP",
"CAP_SETUID",
"CAP_SYS_CHROOT"],"default-ulimits": { "nofile":{ "Name": "nofile", "Hard": 4096, "Soft": 1024 }},  "selinux-enabled": true
}
Aug 21 22:41:39 ip-172-31-71-208.ec2.internal dockerd[1930]: time="2023-08-21T22:41:39.477988449Z" level=info msg="Loading containers: done."
Aug 21 22:41:39 ip-172-31-71-208.ec2.internal dockerd[1930]: time="2023-08-21T22:41:39.650870514Z" level=info msg="Docker daemon" commit=3056208812eb5e792fa99736c9167d1e10f4ab49 graphdriver(s)=overlay2 version=20.10.21
Aug 21 22:41:39 ip-172-31-71-208.ec2.internal dockerd[1930]: time="2023-08-21T22:41:39.651011462Z" level=info msg="Daemon has completed initialization"
Aug 21 22:41:39 ip-172-31-71-208.ec2.internal systemd[1]: Started docker engine.
Aug 21 22:41:39 ip-172-31-71-208.ec2.internal dockerd[1930]: time="2023-08-21T22:41:39.701511993Z" level=info msg="API listen on /run/docker.sock"
```
- [x] aws-ecs-1-nvidia with capabilities, rlimits and container-registry.mirrors
```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" }, "nvidia": { "path": "nvidia-oci" } },"default-capabilities":["CAP_AUDIT_WRITE",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE",
"CAP_FOWNER",
"CAP_FSETID",
"CAP_KILL",
"CAP_MKNOD",
"CAP_NET_BIND_SERVICE",
"CAP_NET_RAW",
"CAP_SETFCAP",
"CAP_SETGID",
"CAP_SETPCAP",
"CAP_SETUID",
"CAP_SYS_CHROOT"],"default-ulimits": { "nofile":{ "Name": "nofile", "Hard": 4096, "Soft": 1024 }},  "registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
```

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

### CODE REVIEW

File: packages/docker-engine/daemon-nvidia-json [hunk 1]
```diff
@@ -6,20 +6,20 @@
...
   {{#if settings.container-registry.mirrors}}
   {{#each settings.container-registry.mirrors}}
-  {{#if (eq registry "docker.io" )}},
-  "registry-mirrors": [{{join_array ", " endpoint}}]
+  {{#if (eq registry "docker.io" )}}
+  "registry-mirrors": [{{join_array ", " endpoint}}],
   {{/if}}
   {{/each}}
   {{/if}}
+  "selinux-enabled": true
```
Comment (yeazelm) [inline packages/docker-engine/daemon-nvidia-json:24]: nit: From this formatting of the template, I didn't expect to see inconsistent newlines in the output. Although its valid JSON, it ends up being difficult to read. This is functionally correct, but I would like to understand how some situations result in a newline and some don't.

**Review by yeazelm — APPROVED** [review]

**Review by arnaldo2792 — APPROVED** [review]

File: packages/docker-engine/daemon-nvidia-json (hunk 1 above)
Comment (vyaghras) [reply packages/docker-engine/daemon-nvidia-json:24]: The ones that we generate dynamically comes in same line

Comment (bcressey) [reply packages/docker-engine/daemon-nvidia-json:24]: It'd be worth trying to fix this to make the output consistent by using the [whitespace control](https://handlebarsjs.com/guide/expressions.html#whitespace-control) `~` more judiciously.

Comment (vyaghras) [reply packages/docker-engine/daemon-nvidia-json:24]: Using ~ (whitespace control) has fixed the indentation. Now config file is as follows:
```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-capabilities": ["CAP_AUDIT_WRITE",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE",
"CAP_FOWNER",
"CAP_FSETID",
"CAP_KILL",
"CAP_MKNOD",
"CAP_NET_BIND_SERVICE",
"CAP_NET_RAW",
"CAP_SETFCAP",
"CAP_SETGID",
"CAP_SETPCAP",
"CAP_SETUID",
"CAP_SYS_CHROOT"],
  "default-ulimits": {
     "nofile":{ "Name": "nofile", "Hard": 4096, "Soft": 1024 }
  },
  "registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
```

**Review by yeazelm — APPROVED** [review]

**Comment by vyaghras** [comment]
Testing all the possible scenarios: 
- [x] Without capabilities, rlimits, registry mirrors
```
bash-5.1# cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "selinux-enabled": true
}
```
- [x] With only capabilities
```
bash-5.1# cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-capabilities": ["CAP_AUDIT_READ",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE"],
  "selinux-enabled": true
}
```
- [x] With Capabilities and rlimits
```
bash-5.1# cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-capabilities": ["CAP_AUDIT_READ",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE"],
  "default-ulimits": {
     "nofile":{ "Name": "nofile", "Hard": 92233720368, "Soft": 40998 }
  },
  "selinux-enabled": true
}
```
- [x] With capabilities, rlimits and registry-mirrors
```
bash-5.1# cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-capabilities": ["CAP_AUDIT_READ",
"CAP_CHOWN",
"CAP_DAC_OVERRIDE"],
  "default-ulimits": {
     "nofile":{ "Name": "nofile", "Hard": 92233720368, "Soft": 40998 }
  },
  "registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
```
- [x] Only rlimits
```
bash-5.1# cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-ulimits": {
     "nofile":{ "Name": "nofile", "Hard": 92233720368, "Soft": 40998 }
  },
  "selinux-enabled": true
}
```
- [x] rlimits and registry-mirrors
```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-ulimits": {
     "nofile":{ "Name": "nofile", "Hard": 92233720368, "Soft": 40998 }
  },
  "registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
```
- [x] Capabilities and registry-mirrors
```
bash-5.1# cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "default-capabilities": ["CAP_AUDIT_READ"],
  "registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
```
- [x] Only registry-mirrors
```
cat etc/docker/daemon.json
{
  "log-driver": "journald",
  "live-restore": true,
  "max-concurrent-downloads": 10,
  "storage-driver": "overlay2",
  "data-root": "/var/lib/docker",
  "default-runtime": "shimpei",
  "runtimes": { "shimpei": { "path": "shimpei" } },
  "registry-mirrors": ["https://registry-0.acme.com", "https://registry-1.acme.com"],
  "selinux-enabled": true
}
```

**Review by bcressey — APPROVED** [review]

**Pull request merged**